miniapps/solvers/schrodinger_flow.hpp

```diff
@@ -0,0 +1,945 @@
+      mfem::forall(ndofs, [=] MFEM_HOST_DEVICE(int n)
+      {
+         const real_t px = X(0, n), py = X(1, n),
+                      pz = DIM == 3 ? X(2, n) : 0.0;
+         const real_t rx = px - o[0], ry = py - o[1], rz = pz - o[2];
+         const real_t z = rx * n0 + ry * n1 + rz * n2;
+         const bool inRange = (rx * rx + ry * ry + rz * rz - z * z) < radius * radius;
+         const bool inLayerP = inRange && (z > 0.0  && z <= (+swirling / 2.0));
+         const bool inLayerM = inRange && (z <= 0.0 && z >= (-swirling / 2.0));
+         real_t alpha = 0.0;
+         if (inLayerP) { alpha = -M_PI * (2.0 * z / swirling - 1.0); }
+         if (inLayerM) { alpha = -M_PI * (2.0 * z / swirling + 1.0); }
+         complex_t psi1(psi1_r(n), psi1_i(n));
+         const complex_t alpha_i(0, alpha);
+         psi1 *= exp(alpha_i);
+         psi1_r(n) = psi1.real(), psi1_i(n) = psi1.imag();
+      });
+   }
+};
+
+/// @brief Crank-Nicolson time solver for the Schrodinger equation.
+template<typename TFiniteElementSpace,
+         typename TSesquilinearForm,
+         typename TComplexGridFunction>
+struct CrankNicolsonBaseSolver
+{
+   ConstantCoefficient one, dthq, mdthq;
+   TSesquilinearForm C_form, R_form;
+   Array<int> no_bc;
+   OperatorHandle C_op, R_op;
+   TComplexGridFunction z;
+   GMRESSolver Cm1;
+
+   // ∂ₜ𝝭 = ½ℏ𝑖∆𝝭
+   // 𝝭ⁿ⁺¹ - 𝝭ⁿ = ¼𝛅t𝑖ℏ(∆𝝭ⁿ⁺¹ + ∆𝝭)
+   // 𝝭ⁿ⁺¹ - ¼𝑖ℏ𝛅t∆𝝭ⁿ⁺¹ = 𝝭ⁿ + ¼𝑖ℏ𝛅t∆𝝭ⁿ
+   // [M + ¼𝑖ℏ𝛅tA]𝝭ⁿ⁺¹ = [M - ¼𝑖ℏ𝛅tA]𝝭ⁿ
+   // C = M + ¼𝑖ℏ𝛅tA, R = M - ¼𝑖ℏ𝛅tA
+   CrankNicolsonBaseSolver(TFiniteElementSpace &fes,
+                           real_t hbar, real_t dt,
+                           std::function<GMRESSolver()> CreateGMRESSolver,
+                           real_t rtol, real_t atol, int maxiter,
+                           int print_level):
+      one(1.0), dthq(dt * hbar / 4.0), mdthq(-dt * hbar / 4.0),
+      C_form(&fes), R_form(&fes),
+      z(&fes),
+      Cm1(CreateGMRESSolver())
+   {
+      // C = M + ¼𝑖ℏ𝛅tA
+      C_form.AddDomainIntegrator(new MassIntegrator(one),
+                                 new DiffusionIntegrator(dthq));
+      C_form.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+      C_form.Assemble();
+      C_form.FormSystemMatrix(no_bc, C_op);
+
+      // R = M - ¼𝑖ℏ𝛅tA
+      R_form.AddDomainIntegrator(new MassIntegrator(one),
+                                 new DiffusionIntegrator(mdthq));
+      R_form.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+      R_form.Assemble();
+      R_form.FormSystemMatrix(no_bc, R_op);
+
+      Cm1.SetPrintLevel(print_level);
+      Cm1.iterative_mode = false;
+      Cm1.SetMaxIter(maxiter);
+      Cm1.SetOperator(*C_op);
+      Cm1.SetRelTol(rtol);
+      Cm1.SetAbsTol(atol);
+   }
+
+   // 𝝭ⁿ⁺¹ = C⁻¹ R 𝝭ⁿ
+   virtual void Mult(TComplexGridFunction &psi) = 0;
+};
+
+/// @brief Class for simulating incompressible Schrodinger flow.
+template<typename TSchrodingerSolver, typename TGridFunction>
+class IncompressibleBaseFlow : private Options
```

Comment:
My apologies if this is a naive question, but what is the advantage of this class inheriting from `Options` as opposed to having a member variable of type `Options`?  I have the same question regarding `VisualizerBase` inheriting from `Options`, but I imagine the answer here will apply there as well.

Reply:
The main advantage of inheriting from `Options` instead of using a member variable is mainly for direct access to members and to keep the code concise.
This can be refactored to composition to use a variable instead if desired.

Reply:
That makes sense.  The use of composition over inheritance is more a general personal preference of mine, and given that I'm not currently a primary user of this miniapp, I don't think anything needs to be refactored... I'm content with the inheritance being private.